### ISSUE TITLE
Be more precise about target CPU features and minimum CPU requirements

### ISSUE THREAD

**Issue by brson** [tikv/tikv#4843]
TL;DR: declare that TiKV runs on westmere-level processors and above. Pass `-march=westmere` to every project we build. Get better optimizations everywhere.

Rationale below.

Today we take some slight consideration about the minimum CPU features required to run TiKV, but there is much more that can usefully be configured to add more optimizations to TiKV.

Today our releases have a slight influence on which CPU features are used to compile RocksDB. We do this by setting PORTABLE=1, and FORCE_SSE42=1 with cmake.

What PORTABLE=1 does to RocksDB is forces it to not pass any `-march` value to gcc. GCC's default -march is x86-64, which looks to be an sse2-level architecture. sse2 was introduced in 2001. Without PORTABLE=0 RocksDB will compile with --march=native, and again, we don't use this for releases.

_We should not be compiling with --march=native, because that makes our minimum architecture dependent on whatever machine the build is running on._

Instead, we should have a clear baseline architecture that TiKV requires, and configure the compiler for all our dependencies to target that architecture.

So for our "portable" release we are targeting a very conservative CPU for RocksDB by setting PORTABLE=1. **BUT**, what FORCE_SSE42 does is explicitly pass `-msse4.2 -mpclmul` to cc.

This is very weird because there is surely no SSE2 level processor that just happens to also have SSE4.2 and MPCLMUL.

What actually changes to the CPU configuration when just those flags are passed is this diff between `gcc -Q --help=target -march=x86-64` (-march can also be left off here because it is the default), and `gcc Q --help=target -msse4.2 -mpclmul`:

```
brian@DESKTOP-UCV672I:~/pingcap/rust-rocksdb/librocksdb_sys$ diff ~/x86-64-target ~/sse42-target                  100c100
<   -mno-sse4                                   [enabled]
---
>   -mno-sse4                                   [disabled]
106c106
<   -mpclmul                                    [disabled]
---
>   -mpclmul                                    [enabled]
109c109
<   -mpopcnt                                    [disabled]
---
>   -mpopcnt                                    [enabled]
133,136c133,136
<   -msse3                                      [disabled]
<   -msse4                                      [disabled]
<   -msse4.1                                    [disabled]
<   -msse4.2                                    [disabled]
---
>   -msse3                                      [enabled]
>   -msse4                                      [enabled]
>   -msse4.1                                    [enabled]
>   -msse4.2                                    [enabled]
140c140
<   -mssse3                                     [disabled]
---
>   -mssse3                                     [enabled]
```

So `FORCE_SSE42` for RocksDB turns on all SSE up to 4.2, popcnt and pclmul.

Aside: SSE4.2 and PCLMUL are only explicitly used in the crc32 implementation, but the flags allow the optimizer to use those instructions as well.

Now, as I said I think this combination of PORTABLE=1 FORCE_SSE42 is very weird. PCLMUL was introduced in the westmere architecture, from 2010. So our minimum architecture is westmere.

Here's a comparison of the features in gcc's x86-64 target vs the westmere target:

```
brian@DESKTOP-UCV672I:~/pingcap/rust-rocksdb/librocksdb_sys$ diff ~/x86-64-target ~/westmere-target
16c16
<   -maes                                       [disabled]
---
>   -maes                                       [enabled]
24c24
<   -march=                                     x86-64
---
>   -march=                                     westmere
53c53
<   -mcx16                                      [disabled]
---
>   -mcx16                                      [enabled]
100c100
<   -mno-sse4                                   [enabled]
---
>   -mno-sse4                                   [disabled]
106c106
<   -mpclmul                                    [disabled]
---
>   -mpclmul                                    [enabled]
109c109
<   -mpopcnt                                    [disabled]
---
>   -mpopcnt                                    [enabled]
125c125
<   -msahf                                      [disabled]
---
>   -msahf                                      [enabled]
133,136c133,136
<   -msse3                                      [disabled]
<   -msse4                                      [disabled]
<   -msse4.1                                    [disabled]
<   -msse4.2                                    [disabled]
---
>   -msse3                                      [enabled]
>   -msse4                                      [enabled]
>   -msse4.1                                    [enabled]
>   -msse4.2                                    [enabled]
140c140
<   -mssse3                                     [disabled]
---
>   -mssse3                                     [enabled]
150c150
<   -mtune=                                     generic
---
>   -mtune=                                     westmere
```

It turns on a few more features. So I think our default "portable" release actually requires a westmere-level processor or above.

The `unportable_release` target does a build where RocksDB is compiled with `-march native`, which I said above, I think is unwise to do. What is the minimum architecture required? It's whatever happens to be running the build machine.

I know I've seen 3.0 release tarballs with two different tikv-server's one portable and the other not. I don't know which makefile targets produce them, but I think they are both wrong when it comes to specifying the CPU architecture. And I suspect we can eliminate one of them by being more specific about minimum target architecture.

So, as I said above, I think our minimum architecture is westmere, a processor from 2010. That seems quite reasonable.

Today we are turning on westmere-like optimizations for RocksDB, but nothing else. I don't know Rust's default -march but I suspect it is the same as gcc/clang - SSE2-level 2001-era x86-64.

So we are losing any opportunity to compile our Rust code with westmere-level optimizations. Ditto for any other C code in the project.

So my proposal is this: declare that TiKV runs on westmere-level processors and above. Pass `-march=westmere` to every project we build. Get better optimizations everywhere.

cc @siddontang @BusyJay @overvenus @nrc

**Comment by brson** [comment]
Some of the additional features added by westmere are pretty interesting: cx16 is 128-bit atomics, aes is hardware AES, popcnt of course is useful for various things, ssse3 is streaming extensions that might be employed in various encoders and decoders. pclmul is "carryless" multiplacation - multiply two 64-bit values get a 128-bit value. sse4.2 also has a crc32 instruction that could be used for something.

From https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html

We might also want to look into "To generate SSE/SSE2 instructions automatically from floating-point code (as opposed to 387 instructions), see -mfpmath=sse." Not sure what rust does by default, or if tikv uses any intense floating point math.

If we did this we might not want to accept `-mtune=westmont` because that literally optimizes for exactly that microarchitecture. If we picked `-march=westmont` we would still probably want to set `-mtune=generic`, or `-mtune=intel`.

**Comment by brson** [comment]
We should probably also do feature detection on startup to avoid mystery illegal instructions during operation for low-class cpus, but not sure if it can be done early enough to avoid hitting westmere-specific code - to do this we could probably put startup code in its own crate that is not -march=westmere.

Another problem here is that std is compiled for -march=x86. This doesn't matter for generics, but some code in std is probably not optimized as much as it could be.

So we could even consider building std ourselves.

**Comment by Hoverbear** [comment]
Related to https://github.com/pingcap/titan/issues/33

**Comment by Hoverbear** [comment]
@brson I support this measure.

@siddontang @zhangjinpeng1987 @BusyJay @nrc PTAL and give your thoughts.

**Comment by siddontang** [comment]
@yiwu-arbug 

Has RocksDB supported this now?

**Comment by yiwu-arbug** [comment]
@siddontang RocksDB cmake script doesn't have similar thing, but I don't think we'll get blocked to add more flags to their cmake script.

**Comment by Hoverbear** [comment]
Are we okay saying our minimum march is for a processor released 9 years ago? https://en.wikipedia.org/wiki/Westmere_(microarchitecture)

I don't see it being a problem, but if we do encounter some problems we could consider distributing a `tikv-legacy` package.

**Comment by nrc** [comment]
This absolutely seems like the right thing to do. We have very beefy recommendations for the minimum spec, so I think requiring a CPU < 10 years old should be fine. Getting those more recent optimisations should make some decent improvements.

It would be interesting to know how many users (and which ones) use the binaries we supply vs building from source. For TiKV specifically (I have no idea about TiDB), I would expect that building from source is an acceptable solution for those who have older processors.

re the native builds, it seems like that should have caused problems already if it were going to? OTOH, we might be leaving performance on the table if people are running with significantly more modern machines than we are building on. Perhaps for paying customers we should offer more fine-grained options for `-march` in our builds?

**Comment by brson** [comment]
I've started a patch for this.

**Comment by brson** [comment]
My patch is https://github.com/tikv/tikv/issues/4843

I'm having some trouble with RocksDB's build system though. It seems to want to control CXXLFAGS in a way that is hard to override. I'm still trying to figure out what it wants.

**Comment by zhangjinpeng1987** [comment]
cc @yiwu-arbug

**Comment by brson** [comment]
I think I figured out the problem I was having with rocks's build system, though I'm still testing it.dd

Rocks's CMakeList.txt generally appends any values it wants to `CMAKE_CXX_FLAGS`, which makes those flags impossible to override (the last one on the command line wins). I'm trying to force the `-march` and `-mtune` variables, but realized that in "portable" mode Rocks's CMakeList.txt doesn't set `-march`, so it's all good. I should be able to override it by turning on portable mode.
